### DISCUSSION TITLE
How can I avoid type hints errors when assigning converted types to a model

### DISCUSSION THREAD

**Discussion by tumma72** [pydantic/pydantic#10450] [Question]
I am not sure the title explains what I am trying to achieve here, and I will try to explain it with a couple of examples. First of all the problem I am trying to solve is to avoid my IDE to thrown type hints (using Ruff on Visual Studio at the moment) errors when doing assignments to Pydantic models attributes with serialized values, which can be imported and converted into a valid model using validators.

For instance, I have a model called SectionNumber:

```python
class SectionNumber(BaseModel):
    """Represents a section number with parts as a list of integers."""

    parts: List[int] = Field(default_factory=list)

    @classmethod
    def from_string(cls, number_string: str) -> "SectionNumber":
        """
        Creates a SectionNumber from a string representation.

        Args:
            number_string: A string representation of the content item number.

        Returns:
            A new SectionNumber instance.
        """
        parts: List[int] = [
            int(part) for part in number_string.split(".") if part.isdigit()
        ]
        return cls(parts=parts)

    @model_validator(mode="before")
    @classmethod
    def validate_parts(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        """
        Validates and converts the 'parts' field if it's a string.

        Args:
            values: A dictionary of field values to validate.

        Returns:
            The validated dictionary of field values.
        """
        if "parts" in values and isinstance(values["parts"], str):
            values["parts"] = cls.from_string(values["parts"]).parts

        return values

    @model_serializer()
    def serialize(self) -> str:
        """Serializes the content item number to a string."""
        return str(self)

    def __str__(self) -> str:
        """Returns a string representation of the content item number."""
        if not self.parts:
            return ""

        return ".".join(map(str, self.parts))
```

I would like to have another model such as Section, using the SectionNumber as an attribute:
```python
class Section(BaseModel):
    number: Optional[SectionNumber] = Field(default_factory=SectionNumber, description="The Section number")
```

Till here all good, what I am trying to find out if possible at all, is the following: I want to be able to create a new Section passing as attribute a number which can be either a string "1.2.4" or an int 2, or an actual `SectionNumber` instance, without having the type hint raising type errors such as:

```
Argument of type "Literal['1.1.2']" cannot be assigned to parameter "number" of type "ContentItemNumber | None" in function "__init__"
  Type "Literal['1.1.2']" is not assignable to type "ContentItemNumber | None"
    "Literal['1.1.2']" is not assignable to "ContentItemNumber"
    "Literal['1.1.2']" is not assignable to "None"
```

Is there any way in which I could achieve this? Perhaps using an Annotated type? Thanks in advance for your help.

**Comment by RamiAwar** [comment]
Sorry if this is unhelpful, but from a high level perspective this looks like something you shouldn't be doing in the first place, hence the difficulty! I might have misunderstood the question, but here's what I'm hearing:

If you really want to go down this path and instantiate a `Section` from a string, a `SectionNumber`, or an `int`, then maybe having an if statement is best:

```python
def create_section(x: str | int | SectionNumber) -> Section:
    if isinstance(x, str):
        return Section(number=SectionNumber.from_string(x))
    elif isinstance(x, int):
        # do from_int
    else:
        return Section(number=x)
```

**Comment by tumma72** [comment]
Thanks for the quick reply,
> Sorry if this is unhelpful, but from a high level perspective this looks like something you shouldn't be doing in the first place, hence the difficulty!
> 
I figured that much, 
> If you really want to go down this path and instantiate a Section from a string, a SectionNumber, or an int, then maybe having an if statement is best:
> 
> def create_section(x: str | int | SectionNumber) -> Section:
>     if isinstance(x, str):
>         return Section(number=SectionNumber.from_string(x))
>     elif isinstance(x, int):
>         # do from_int
>     else:
>         return Section(number=x)
A factory method is a good pattern to solve the problem and I will probably end up doing this, which will require change the code in a lot of places… but it would be type cleaner at least. My question was aiming at understanding of something like this would be possible:
1. Declare an attribute with a type like SectionNumber
2. Create a series of validators which take a different type as input and convert it to the declared type
3. Find a way to ensure that from an IDE perspective the attribute responds to the declared type when is accessed and it would allow assignments from all types declared as valid inputs from the validators

I was thinking there is a combination of magic declarations using TypeVar and Annotated but obviously there isn’t. 

So thanks for your straightforward reply

Ciao
ANdreaT

**Reply by RamiAwar** [reply]
Ohh now I see what you're trying to do. You could probably do it with validators. Then the validator applies the factory methods internally.

But typing wise you'll have to type it as `SectionNumber | str | int`. That doesn't look too bad at first sight, it is the expected type and users of your code would be using it correctly.

**Reply by tumma72** [reply]
Right and that is what I was trying to avoid, normally I'll do this with a property setter or a descriptor, but those don't work with Pydantic. I want the code completion and type checking to work as if the actual type would always be a `SectionNumber` but at the same time being allowed to assign the number using a string or an int, that's it. 

The funny thing is that if I wouldn't type annotate the variable at all it would just work, without completions or perhaps with a clever enough IDE introspection it would eventually figure that out. 

Anyways, I'll end up with factory methods `from_str` and `from_int` and call those from a `model_validator`.

**Reply by tumma72** [reply]
I might have found something in the Pydantic types code, which goes in the right direction:

```python
from typing import Any
from typing_extensions import Annotated

from pydantic import BaseModel, GetPydanticSchema

HandleAsAny = GetPydanticSchema(lambda _s, h: h(Any))

class Model(BaseModel):
    x: Annotated[int, HandleAsAny]  # pydantic sees `x: Any`

print(repr(Model(x='abc').x))
#> 'abc'
```

This could be used as example to have:

```python
from typing_extensions import Annotated

from pydantic import BaseModel, GetPydanticSchema

from section_number import SectionNumber

HandleAsSectionNumber = GetPydanticSchema(lambda _s, h: h(SectionNumber))

class Section(BaseModel):
    number: Annotated[int, str, HandleAsSectionNumber]
```

Would that work? Or am I missing something?

**Reply by RamiAwar** [reply]
Why don't you want to set the type as `str | int | SectionNumber`? That's actually the type, why do you want to 'hide' it?

**Reply by tumma72** [reply]
Because it isn't! The type is a `SectionNumber` when you access the attribute as `Section.number` it will always behave like a `SectionNumber` but you can create one by passing an `int` or a `str` as well. I just want to avoid that the developers using the framework will have to import the `SectionNumber` and create an explicit instance all the times. As mentioned in a previous post I would normally do this with a property setter or a descriptor but with Pydantic I need to use a model_validator to convert `Section.number` to a `SectionNumber`. Does this make sense?

**Reply by RamiAwar** [reply]
Okay I see what you mean. When creating a SectionNumber, the type is indeed str | int | SectionNumber but if you're simply getting the attribute it'll always be SectionNumber.

**Comment by Viicos** [comment]
This is a limitation of the Python type system, see: https://github.com/pydantic/pydantic/issues/8616#issuecomment-1907572927.
